### PR TITLE
Detect and deduplicate read-only resource access.

### PR DESCRIPTION
**Pull request by csyonghe**
Mark all read-only resource intrinsics as `[__readNone]`.

Deduplicate `load`s of `ConstantBuffer` and `ParameterBlock`.

### CODE REVIEW

File: source/slang/core.meta.slang [hunk 1]
```diff
@@ -1189,6 +1189,7 @@ for (int tt = 0; tt < kBaseTextureTypeCount; ++tt)
                     continue;
                 }
 
+                bool isReadOnly = (access == SLANG_RESOURCE_ACCESS_READ);
```
Comment (jsmall-nvidia) [inline source/slang/core.meta.slang:1192]: Is there a possible issue with aliasing? The [__readNone] attribute in the context of a resource method would mean the methods result is entirely determined by its inputs, with the not unreasonable assumption that if it's a read only resource the resources contents can't change.

But if it is possible to mutate the contents of a read only resource, via a writable alias on some aspect of the resource then that seems problematic.

Comment (csyonghe) [reply source/slang/core.meta.slang:1192]: The HLSL compiler assumes no aliasing by default, and would require the user to pass in a specific flag to inform the compiler that there is aliasing and should not perform any optimizations.

We don't have such a flag yet in Slang, and we should probably introduce it.

Comment (csyonghe) [reply source/slang/core.meta.slang:1192]: Given that our dxc backend never generates the dxc flag to enable aliasing, I guess it is safe for us to assume there is no such uses out there in our user codebase. But we should definitely inform people about this change.

Comment (jsmall-nvidia) [reply source/slang/core.meta.slang:1192]: Ok sounds good.

**Review by jsmall-nvidia — DISMISSED** [review]

**Pull request merged**